File: python/P36.py
class Solution(object):
    def isValidSudoku(self, board):
        m = len(board)
        n = len(board[0])
        for i in range(m):
            for j in range(n):
                if board[i][j] != '.' and not self.isvalid(i, j, board):
                    return False
        return True

    def isvalid(self, x, y, board):
        m = len(board)
        n = len(board[0])
        for i in range(m):
            if board[i][y] == board[x][y] and i != x:
                return False
        for i in range(n):
            if board[x][y] == board[x][i] and i != y:
                return False
        for i in range(3):
            for j in range(3):
                if board[x/3*3+i][y/3*3+j] == board[x][y] and x/3*3+i != x and y/3*3+j != y:
                    return False
        return True

File: java/P36.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        if (board == null || board.length == 0) return false;
        for (int i = 0; i < 9; i++) {
            HashSet<Character> row = new HashSet<Character>();
            HashSet<Character> col = new HashSet<Character>();
            HashSet<Character> square = new HashSet<Character>();
            for (int j = 0; j < 9; j++) {
                if (board[i][j] != '.' && !row.add(board[i][j])) return false;
                if (board[j][i] != '.' && !col.add(board[j][i])) return false;
                int rowindex = 3*(i/3);
                int colindex = 3*(i%3);
                if (board[rowindex+j/3][colindex+j%3] != '.' && !square.add(board[rowindex+j/3][colindex+j%3])) return false;
            }
        }
        return true;
    }
}